proselint/checks/__init__.py

```diff
@@ -1 +1,394 @@
+###############################################################################
+
+
+def consistency_check(
+    text: str,
+    word_pairs: list,
+    err: str,
+    msg: str,
+    offset: int = 0,
+) -> list[ResultCheck]:
+    """Build a consistency checker for the given word_pairs."""
+    results = []
+
+    for w in word_pairs:
+        matches = [
+            list(re.finditer(w[0], text)),
+            list(re.finditer(w[1], text)),
+        ]
+
+        if len(matches[0]) > 0 and len(matches[1]) > 0:
+            idx_minority = len(matches[0]) > len(matches[1])
+
+            results += [
+                (
+                    m.start() + offset,
+                    m.end() + offset,
+                    err,
+                    msg.format(w[not idx_minority], m.group(0)),
+                    w[not idx_minority],
+                )
+                for m in matches[idx_minority]
+            ]
+
+    return results
+
+
+def preferred_forms_check(  # noqa: PLR0913, PLR0917
+    text: str,
+    items: list,
+    err: str,
+    msg: str,
+    ignore_case: bool = True,
+    offset: int = 0,
+) -> list[ResultCheck]:
+    """Build a checker that suggests the preferred form."""
+    flags = re.IGNORECASE if ignore_case else 0
+    regex = r"[\W^]{}[\W$]"  # TODO: replace with enum below?
+
+    return [
+        (
+            m.start() + 1 + offset,
+            m.end() + offset,
+            err,
+            msg.format(item[0], m.group(0).strip()),
+            item[0],
+        )
+        for item in items
+        for r in item[1]
+        for m in re.finditer(regex.format(r), text, flags=flags)
+    ]
+    # TODO: can we speed up str.format() ?
+    #       fast-string? or do padding already in checks
+
+
+# PADDINGS, add more if needed
+class Pd(str, Enum):
+    disabled = r"{}"
+    # choose for checks with custom regex
+    whitespace = r"\s{}\s"
+    # -> req whitespace around (no punctuation!)
+    sep_in_txt = r"(?:^|\W){}[\W$]"
+    # req non-text character around
+    # -> finds item as long it is surrounded by any non-word character:
+    #       - whitespace
+    #       - punctuation
+    #       - newline ...
+
+
+def existence_check(  # noqa: PLR0913, PLR0917
+    text: str,
+    re_items: list,
+    err: str,
+    msg: str,
+    ignore_case: bool = True,
+    string: bool = False,  # todo: why not default on?
+    offset: int = 0,  # todo: some checks set this strangely
+    padding: Pd = Pd.sep_in_txt,
+    dotall: bool = False,
+    excluded_topics: Optional[list] = None,
+    exceptions=(),
+) -> list[ResultCheck]:
+    """Build a checker that prohibits certain words or phrases."""
+    flags = 0
+    if ignore_case:
+        flags |= re.IGNORECASE
+    if string:
+        flags |= re.UNICODE
+    if dotall:
+        flags |= re.DOTALL
+
+    errors: list[ResultCheck] = []
+
+    # If the topic of the text is in the excluded list, return immediately.
```

Comment:
Topics are hard to maintain, and if I remember correctly, the list is very limited and unfinished. I would lean towards removing this with a possibility of later reconsideration if we can figure out a good system for it, honestly.

Reply:
the idea is good - implementation is bad. i can think of some more checks that would benefit from it. currently it's only used by `dates/check_decade_apostrophes_short`. we can drop the feature for now but it's also minimal overhead

Reply:
What other checks would benefit from this kind of topic detection?